assets/js/src/tracking.js

```diff
@@ -0,0 +1,219 @@
+ *
+ * @type {number}
+ */
+let currentStep = -1;
+
+/**
+ * Tracks view_item_list event
+ *
+ * @param {Object} params The function params
+ * @param {Array} params.products The products to track
+ * @param {string} [params.listName] The name of the list in which the item was presented to the user.
+ */
+export const trackListProducts = ( {
+	products,
+	listName = __( 'Product List', 'woocommerce-google-analytics-integration' ),
+} ) => {
+	trackEvent( 'view_item_list', {
+		event_category: 'engagement',
+		event_label: __(
+			'Viewing products',
+			'woocommerce-google-analytics-integration'
+		),
+		items: products.map( ( product, index ) => ( {
+			...getProductImpressionObject( product, listName ),
+			list_position: index + 1,
+		} ) ),
+	} );
+};
+
+/**
+ * Tracks add_to_cart event
+ *
+ * @param {Object} params The function params
+ * @param {Array} params.product The product to track
+ * @param {number} [params.quantity=1] The quantity of that product in the cart.
+ */
+export const trackAddToCart = ( { product, quantity = 1 } ) => {
+	trackEvent( 'add_to_cart', {
+		event_category: 'ecommerce',
+		event_label: __(
+			'Add to Cart',
+			'woocommerce-google-analytics-integration'
+		),
+		items: [ getProductFieldObject( product, quantity ) ],
+	} );
+};
+
+/**
+ * Tracks remove_from_cart event
+ *
+ * @param {Object} params The function params
+ * @param {Array} params.product The product to track
+ * @param {number} [params.quantity=1] The quantity of that product in the cart.
+ */
+export const trackRemoveCartItem = ( { product, quantity = 1 } ) => {
+	trackEvent( 'remove_from_cart', {
+		event_category: 'ecommerce',
+		event_label: __(
+			'Remove Cart Item',
+			'woocommerce-google-analytics-integration'
+		),
+		items: [ getProductFieldObject( product, quantity ) ],
+	} );
+};
+
+/**
+ * Tracks change_cart_quantity event
+ *
+ * @param {Object} params The function params
+ * @param {Array} params.product The product to track
+ * @param {number} [params.quantity=1] The quantity of that product in the cart.
+ */
+export const trackChangeCartItemQuantity = ( { product, quantity = 1 } ) => {
+	trackEvent( 'change_cart_quantity', {
+		event_category: 'ecommerce',
+		event_label: __(
+			'Change Cart Item Quantity',
+			'woocommerce-google-analytics-integration'
+		),
+		items: [ getProductFieldObject( product, quantity ) ],
+	} );
+};
+
+/**
+ * Track a begin_checkout and checkout_progress event
+ * Notice calling this will set the current checkout step as the step provided in the parameter.
+ *
+ * @param {number} step The checkout step for to track
+ * @return {(function( { storeCart: Object } ): void)} A callable receiving the cart to track the checkout event.
+ */
+export const trackCheckoutStep =
+	( step ) =>
+	( { storeCart } ) => {
+		if ( currentStep === step ) {
+			return;
+		}
+
+		trackEvent( step === 0 ? 'begin_checkout' : 'checkout_progress', {
```

Comment:
GA4 drops `checkout_progress` completely and doesn't include `checkout_step` option